pkg/resources/common/common.go

```diff
@@ -14,3 +20,82 @@ func ConvertAuthnExtras(extra map[string]authnv1.ExtraValue) map[string]authzv1.
 	}
 	return result
 }
+
+// ValidateLabel checks if a user is removing or modifying a label. If the label is newly added, return false.
+func IsModifyingLabel(oldLabels, newLabels map[string]string, label string) bool {
+	var oldValue, newValue string
+	var oldLabelExists, newLabelExists bool
+	if oldLabels == nil {
+		oldLabelExists = false
+	} else {
+		oldValue, oldLabelExists = oldLabels[label]
+	}
+	if newLabels == nil {
+		newLabelExists = false
+	} else {
+		newValue, newLabelExists = newLabels[label]
+	}
+
+	if !oldLabelExists {
+		return false
+	}
+	if oldLabelExists && !newLabelExists {
+		return true
+	}
+	if oldValue != newValue {
+		return true
+	}
+
+	return false
+}
+
+// CachedVerbChecker is used for caching if a request for a non-namespaced gvr with specified name has the given overrideVerb.
+// This is meant to eliminate the need to perform multiple calls to the provided SubjectAccessReview for the overrideVerb.
+// Each CachedVerbChecker is unique to the initial set up. If the caller needs to change what it is checking
+// (different verb, resource name, resource type) a new CachedVerbChecker must be created.
+// A CachedVerbChecker should not be shared between admitters. Each admitter must request a new CachedVerbChecker.
+// Additionally, the CachedVerbChecker should not be shared between requests, even for the same admitter.
+type CachedVerbChecker struct {
+	request            *admission.Request
+	name               string
+	sar                authorizationv1.SubjectAccessReviewInterface
+	gvr                schema.GroupVersionResource
+	overrideVerb       string
+	hasVerbBeenChecked bool
+	hasVerb            bool
+}
+
+// NewCachedVerbChecker creates a new CachedVerbChecker
+func NewCachedVerbChecker(req *admission.Request, name string, sar authorizationv1.SubjectAccessReviewInterface, gvr schema.GroupVersionResource, verb string) *CachedVerbChecker {
+	return &CachedVerbChecker{
+		request:            req,
+		name:               name,
+		sar:                sar,
+		gvr:                gvr,
+		overrideVerb:       verb,
+		hasVerbBeenChecked: false,
+		hasVerb:            false,
+	}
+}
+
+// IsRulesAllowed checks if the request has permissions to create the rules provided. Returns nil if the rules are allowed.
+func (c *CachedVerbChecker) IsRulesAllowed(rules []v1.PolicyRule, resolver validation.AuthorizationRuleResolver, namespace string) error {
+	err := auth.ConfirmNoEscalation(c.request, rules, namespace, resolver)
+	// Check for the overrideVerb
+	if err != nil {
+		if c.HasVerb() {
+			return nil
+		}
+	}
+	return err
+}
+
+// HasVerb returns if the request has the overrideVerb. Only checks the request the first time called, after that it returns the cached value.
+func (c *CachedVerbChecker) HasVerb() bool {
+	if c.hasVerbBeenChecked {
+		return c.hasVerb
+	}
+	c.hasVerb, _ = auth.RequestUserHasVerb(c.request, c.gvr, c.sar, c.overrideVerb, c.name, "")
```

Comment:
Nit: Generally speaking you never want to ignore an error. In this case, I think we can allow it since the calling function is one that we wrote and the functionality is pretty clear on the cases where it returns an error, but it's definitely something that we should avoid. I'd go with something like:

```suggestion
	c.hasVerb, err = auth.RequestUserHasVerb(c.request, c.gvr, c.sar, c.overrideVerb, c.name, "")
        if err != nil{
                // Ideally this wouldn't log and would return up to the calling function/admitters, but this would be acceptable
	        logrus.Errorf("got an error when checking if a user has a verb %s", err)
	        return false
	}
```

This also has the disadvantage of "eating the error" so if there's something wrong with the SAR logic we will have no way of telling.

Reply:
I added a log and made a test for it